Add support for Actinius Icarus SoM Development Kit

alextsam opened the pull request:
This adds support for the Actinius Icarus SoM DK (www.actinius.com/icarus-som-dk) and patches other Actinius boards for silencing a build-time warning.

Signed-off-by: Alex Tsamakos <alex@actinius.com>

alextsam commented:
@gmarull @carlescufi can this make it into Zephyr before the upcoming release?

gmarull commented on boards/arm/actinius_icarus/board.c:
```diff
@@ -5,82 +5,8 @@
-
-/* Needs to happen after GPIO driver init */
-SYS_INIT(board_actinius_icarus_init, POST_KERNEL, 99);
+/* EMPTY */
```
please, delete empty files.

gmarull replied:
you should be able to know at build time wether you need board.c or not (in CMake)

alextsam replied:
@gmarull the condition in CMake is already there, this is to avoid the warning when CMake does not have anything to do for `secure`. This is the warning that comes up if we don't include this "trick":
```
CMake Warning at C:/ncs/v2.1.0/zephyr/CMakeLists.txt:831 (message):
  No SOURCES given to Zephyr library: boards__arm__actinius_icarus

  Excluding target from build.
```
 
![image](https://user-images.githubusercontent.com/4224959/203584084-b3e5f5b1-0e4f-4a68-ab9f-2ff62a53e8ca.png)

alextsam replied:
This is addressed with the helper lib, although the warning remains

gmarull commented on boards/arm/actinius_icarus/board_ns.c:
```diff
@@ -0,0 +1,86 @@
+}
+
+/* Needs to happen after GPIO driver init */
+SYS_INIT(board_actinius_icarus_init, POST_KERNEL, 99);
```
nit: avoid hardcoding levels, create a Kconfig option

alextsam replied:
This was done on par with https://github.com/zephyrproject-rtos/zephyr/pull/33740 and the value should be always after GPIO init (thus 99 as it's the last one). IMO it shouldn't be left configurable; what do you think?

alextsam replied:
This is addressed with the helper lib

gmarull commented on boards/arm/actinius_icarus_bee/board.c:
```diff
@@ -5,35 +5,8 @@
-
-/* Needs to happen after GPIO driver init */
-SYS_INIT(board_actinius_icarus_bee_init, POST_KERNEL, 99);
+/* EMPTY */
```
ditto

alextsam replied:
This is addressed with the helper lib

gmarull commented on boards/arm/actinius_icarus_som_dk/actinius_icarus_som_dk_common.dts:
```diff
@@ -0,0 +1,227 @@
+		compatible = "st,lis2dh12", "st,lis2dh";
+		reg = <0x19>;
+		irq-gpios = <&gpio0 29 GPIO_ACTIVE_HIGH>,
+			  <&gpio0 28 GPIO_ACTIVE_HIGH>;
```
nit: fix alignment

gmarull commented on boards/arm/actinius_icarus_som_dk/board.c:
```diff
@@ -0,0 +1,12 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(board_control, CONFIG_BOARD_ICARUS_SOM_DK_LOG_LEVEL);
+
+/* EMPTY */
```
ditto

alextsam replied:
This is addressed with the helper lib

mbolivar-nordic commented on boards/arm/actinius_icarus_som_dk/board_ns.c:
```diff
@@ -0,0 +1,39 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(board_control, CONFIG_BOARD_ICARUS_SOM_DK_LOG_LEVEL);
+
+static int board_actinius_icarus_som_dk_init(const struct device *dev)
+{
+	const struct gpio_dt_spec sim =
+		GPIO_DT_SPEC_GET(DT_NODELABEL(sim_select), sim_gpios);
+
+	ARG_UNUSED(dev);
+
+	if (!device_is_ready(sim.port)) {
+		LOG_ERR("The SIM Select Pin port is not ready");
+
+		return -ENODEV;
+	}
+
+	if (DT_ENUM_IDX(DT_NODELABEL(sim_select), sim) == 0) {
+		(void)gpio_pin_configure_dt(&sim, GPIO_OUTPUT_HIGH);
+		LOG_INF("eSIM is selected");
+	} else {
+		(void)gpio_pin_configure_dt(&sim, GPIO_OUTPUT_LOW);
+		LOG_INF("External SIM is selected");
+	}
+
+	return 0;
```
So this is a library that is shared by multiple boards by the same vendor, right?

I created `boards/common` when I was adding the runners infrastructure with the intention that more common board files could go here. Maybe `boards/common/actinius/` could hold a helper library that you could stash this in?

It's interesting because this is something that hardware model v2 (https://github.com/zephyrproject-rtos/zephyr/issues/51831) seems to support naturally, since we would have `boards/actinius/actinius_icarus_som_dk` alongside `boards/actinius/actinius_some_other_board`, and you could more easily add `boards/actinius/helper_lib`.

alextsam replied:
Should I make a helper lib in `boards/common` or shall we wait to do this directly with hardware model v2 later?

mbolivar-nordic replied:
> Should I make a helper lib in `boards/common` or shall we wait to do this directly with hardware model v2 later?

If you want this merged anytime soon, I would go with a new lib in boards/common

alextsam replied:
Alright, I added a helper library in boards/common and changed the som dk to use it. Please let me know if it looks good to you and I will then squash and add a commit for changing the other boards as well asap.
I have a couple of questions:
- Is there a way to avoid having to explicitly source the lib Kconfig in Kconfig.defconfig?
- The dts bindings stay with the board, right?
- CMake complains

```
CMake Warning at ../../../CMakeLists.txt:831 (message):
  No SOURCES given to Zephyr library: boards__arm__actinius_icarus_som_dk

  Excluding target from build.
```
Any suggestions for that?
Thank you.

carlescufi commented:
@alextsam please address review comments

mbolivar-nordic commented on boards/common/actinius/actinius_board_common.c:
```diff
@@ -0,0 +1,93 @@
+#define SIM_SELECT_NODE DT_NODELABEL(sim_select)
+#define CHARGER_ENABLE_NODE DT_NODELABEL(charger_enable)
+
+#if IS_ENABLED(CONFIG_ACTINIUS_BOARD_HAS_SIM_SELECT)
```
just use `#ifdef CONFIG_...`, here and elsewhere. You only need IS_ENABLED when you need either 1 or 0 outside of an ifdef

alextsam replied:
fixed, thanks!

alextsam committed:
boards: actinius_*: change all boards to use common/actinius lib

This changes all Actinius boards to use the helper lib in
`boards/common/actinius` for setting up common init values
such as SIM Select and Charger Enable.

Signed-off-by: Alex Tsamakos <alex@actinius.com>

gmarull commented on boards/arm/actinius_icarus_som_dk/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+	zephyr_library_sources(board_ns.c)
+else()
+	zephyr_library_sources(board.c)
+endif()
```
`board.c` is an empty file, so please delete it and just do

```
if(CONFIG_BOARD_ACTINIUS_ICARUS_SOM_DK_NS)
  zephyr_library()
	zephyr_library_sources(board_ns.c)
endif()
```

alextsam replied:
This whole block is not used after the introduction of the common lib. Should I squash the 2 commits so that the som dk and the common lib come in in a single commit?

gmarull replied:
I think so, yes

alextsam replied:
Done!

gmarull commented on boards/common/actinius/Kconfig:
```diff
@@ -0,0 +1,24 @@
+	help
+	  Sets the init priority for the Board Control module.
+
+config ACTINIUS_BOARD_HAS_SIM_SELECT
+	bool
+	help
+	  Should be selected if an Actinius board provides
+	  a choice for internal or external SIM.
+
+config ACTINIUS_BOARD_HAS_CHARGER_ENABLE
+	bool
+	help
+	  Should be selected if an Actinius board provides
+	  a choice for disabling the charger.
```
this is redundant: can be determined using DT API. E.g. `DT_HAS_COMPAT_STATUS_OKAY(actinius_sim_select)`

alextsam replied:
alright, I dropped these two configs, thanks!

gmarull commented on boards/arm/actinius_icarus_som_dk/arduino_connector.dtsi:
```diff
@@ -0,0 +1,51 @@
+	arduino_adc: analog_connector {
+		compatible = "arduino,uno-adc";
+		#io-channel-cells = <1>;
+		io-channel-map = <0 &adc 2>,	/* A0 = P0.15 = AIN2 */
+				 <1 &adc 3>,	        /* A1 = P0.16 = AIN3 */
+				 <2 &adc 4>,	        /* A2 = P0.17 = AIN4 */
+				 <3 &adc 5>,	        /* A3 = P0.18 = AIN5 */
+				 <4 &adc 6>,	        /* A4 = P0.19 = AIN6 */
+				 <5 &adc 7>;	        /* A5 = P0.20 = AIN7 */
+	};
```
please align properly here

alextsam replied:
done

gmarull commented on boards/arm/actinius_icarus_som_dk/actinius_icarus_som_dk_common.dts:
```diff
@@ -0,0 +1,227 @@
+	lis2dh12_accel: lis2dh12-accel@19 {
+		compatible = "st,lis2dh12", "st,lis2dh";
+		reg = <0x19>;
+		irq-gpios = <&gpio0 29 GPIO_ACTIVE_HIGH>,
+					<&gpio0 28 GPIO_ACTIVE_HIGH>;
```
align properly

alextsam replied:
I improved the indentation, but these cannot align completely without either using spaces or by pushing the gpio of the first line further, which in my opinion looks weird

gmarull commented on boards/arm/actinius_icarus_som_dk/actinius_icarus_som_dk_common.dts:
```diff
@@ -0,0 +1,227 @@
+		spi-max-frequency = <4000000>;
+
+		chain-length = <1>;
+		color-mapping = <LED_COLOR_ID_GREEN
+						LED_COLOR_ID_RED
+						LED_COLOR_ID_BLUE>;
```
alignment

alextsam replied:
done

gmarull commented on boards/common/actinius/actinius_board_common.c:
```diff
@@ -0,0 +1,93 @@
+/* Needs to happen after GPIO driver init */
+SYS_INIT(actinius_board_init,
+		POST_KERNEL,
+		CONFIG_ACTINIUS_BOARD_CONTROL_INIT_PRIORITY);
```
SYS_INIT entry is sometimes optional, make all block conditional to `DT_HAS_COMPAT_STATUS_OKAY(actinius_sim_select) || DT_HAS_COMPAT_STATUS_OKAY(actinius_charger_enable)`

alextsam replied:
done

alextsam committed:
boards: arm: add support for the Actinius Icarus SoM DK w/ helper lib

This adds support for the nRF9160-based Icarus SoM DK
(development kit) from Actinius and a common library for
board init code that is common to multiple boards
from Actinius.

Signed-off-by: Alex Tsamakos <alex@actinius.com>

mbolivar-nordic commented on boards/common/actinius/actinius_board_common.c:
```diff
@@ -0,0 +1,98 @@
+#endif
+
+#if DT_HAS_COMPAT_STATUS_OKAY(actinius_charger_enable)
+	result = actinius_board_set_charger_enable();
```
Won't a success here overwrite an error on line 75?

alextsam replied:
Fixed the commit log, thanks, I missed this earlier.

Regarding the error handling, basically it's relying on logging here; maybe we should even remove the `return` as this function is only used with SYS_INIT and -please correct me if I am wrong- there is no way to get the return value from the kernel.

mbolivar-nordic replied:
> maybe we should even remove the return

No objections from my side if it's OK to proceed despite errors on your hardware

mbolivar-nordic reviewed (commented):
Can you fix the commit logs? Looks like you have some extra signed-off-by lines due to some squashing. One question about error handling -- is this really what you mean?

Generally looks fine to me, thanks for your persistence!
